presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveParquetDereferencePushDown.java

```diff
@@ -0,0 +1,407 @@
+            }
+
+            Map<RowExpression, Subfield> dereferenceToNestedColumnMap = extractDereferences(
+                    session,
+                    rowExpressionService.getExpressionOptimizer(),
+                    new HashSet<>(project.getAssignments().getExpressions()));
+            if (dereferenceToNestedColumnMap.isEmpty()) {
+                return visitPlan(project, context);
+            }
+
+            Map<String, HiveColumnHandle> regularHiveColumnHandles = tableScan.getAssignments().values().stream()
+                    .map(columnHandle -> (HiveColumnHandle) columnHandle)
+                    .collect(toMap(HiveColumnHandle::getName, identity()));
+
+            List<VariableReferenceExpression> newOutputVariables = new ArrayList<>(tableScan.getOutputVariables());
+            Map<VariableReferenceExpression, ColumnHandle> newAssignments = new HashMap<>(tableScan.getAssignments());
+
+            Map<RowExpression, VariableReferenceExpression> dereferenceToVariableMap = new HashMap<>();
+
+            for (Map.Entry<RowExpression, Subfield> dereference : dereferenceToNestedColumnMap.entrySet()) {
+                Subfield nestedColumn = dereference.getValue();
+                RowExpression dereferenceExpression = dereference.getKey();
+
+                // Find the nested column Hive Type
+                HiveColumnHandle regularColumnHandle = regularHiveColumnHandles.get(nestedColumn.getRootName());
+                if (regularColumnHandle == null) {
+                    throw new IllegalArgumentException("nested column [" + nestedColumn + "]'s base column " + nestedColumn.getRootName() + " is not present in table scan output");
+                }
+
+                Optional<HiveType> nestedColumnHiveType = regularHiveColumnHandles.get(nestedColumn.getRootName())
+                        .getHiveType()
+                        .findChildType(
+                                nestedColumn.getPath().stream()
+                                        .map(p -> ((Subfield.NestedField) p).getName())
+                                        .collect(Collectors.toList()));
+
+                if (!nestedColumnHiveType.isPresent()) {
+                    throw new IllegalArgumentException("nested column [" + nestedColumn + "] type is not present in Hive column type");
+                }
+
+                String pushdownColumnName = pushdownColumnNameForSubfield(nestedColumn);
+                // Create column handle for nested column
+                HiveColumnHandle nestedColumnHandle = new HiveColumnHandle(
+                        pushdownColumnName,
+                        nestedColumnHiveType.get(),
+                        dereferenceExpression.getType().getTypeSignature(),
+                        -1,
+                        REGULAR,
+                        Optional.of("nested column pushdown"),
+                        ImmutableList.of(),
+                        Optional.of(nestedColumn));
```

Comment:
If we have this to be "SYNTHESIZED", then we don't need two subfields variables here. We just use one. The SYNTHESIZED one is to denote parquet pushdown and the regular one for selective page source.